core/trino-main/src/main/java/io/trino/sql/planner/NodePartitioningManager.java

```diff
@@ -260,16 +263,27 @@ public BucketNodeMap getBucketNodeMap(Session session, PartitioningHandle partit
      */
     private int getDefaultBucketCount(Session session, PartitioningHandle partitioningHandle)
     {
-        // The goal is to have enough buckets to evenly distribute them across tasks created.  The simplest way to do that is to have exactly as many buckets as there are tasks;
-        // then each task gets one bucket.  So let's return the total number of tasks here.  This number, unfortunately, is hard to know precisely because of the variety of ways in
-        // which schedulers create tasks.  We therefore have to estimate.
+        // The default bucket count is used by both remote and local exchanges to assign buckets to nodes and drivers. The goal is to have enough
+        // buckets to evenly distribute them across tasks or drivers. If number of buckets is too low, then some tasks or drivers will be idle.
+        // Excessive number of buckets would make bucket lists or arrays too large.
```

Comment:
I'm not sure a thousand array elements here or there is a big issue.  To me the primary reason to limit the count is that too many buckets can also cause skew.  For instance, if there are 1.5x as many buckets as nodes, then half the nodes will have half the load.

Reply:
> To me the primary reason to limit the count is that too many buckets can also cause skew. For instance, if there are 1.5x as many buckets as nodes, then half the nodes will have half the load.


The data will be skewed even if you have 1x buckets as nodes. You just don't see that because buckets are glued together. Ideally you have as many buckets as possible values, but we don't know all possible values (or it might be super large), so we have to settle with something smaller.

Reply:
> Skewed buckets will just be glued together.

That seems like a separate problem that we ought to solve independently.  We should be assigning buckets to tasks uniformly regardless of the bucket count, right?

>  Ideally you have as many buckets as possible values, but we don't know all possible values (or it might be super large), so we have to settle with something smaller.

Right, but there are many somethings that are smaller.  I figure that the ideal bucket count is K times the task count, where K is a positive integer.  So I made K=1.

Even if we know the task count precisely and the bucket->task mapping is uniform, we'll still have skew when K is not an integer.  And the `max` calculation below makes that more likely -- it may pick `localBucketCount` when the caller is in fact calculating for the `remoteBucketCount` case.  Your https://github.com/trinodb/trino/pull/22000#discussion_r1603663883 points to one branch of an `if`; the value appropriate for one branch may be wrong for another.  That's why I'm asking if we should split the two calculations.

Reply:
> That seems like a separate problem that we ought to solve independently. We should be assigning buckets to tasks uniformly regardless of the bucket count, right?

I'm not sure what you mean, but a single bucket to partition mapping won't be different if you have 1x, 1.5x or 3x `bucket/node` ratio.

>  I figure that the ideal bucket count is K times the task count, where K is a positive integer. So I made K=1.

The ideal bucket count is actual number of buckets that bucket function could produce. However, this is impractical if bucket count is "infinite", but it also doesn't hurt if bucket count used by engine is higher than actual bucket count.

The problem is when bucket count used by engine is lower than actual bucket count, because then engine has less granularity to work with when shuffling data around. For scaling of writers this also means that single "virtual" bucket would be writing to more than a single real bucket, which means data is fragmented more when scaling of writers happens.

Reply:
Alternatively, we can force the caller to pass the exact number of tasks to this method, be that partition count, buffer count, node count, or whatever.

Reply:
The caller might not know the partition count.

Reply:
OK.